File: packages/mobile/native/android/java/com/bendyline/gezel/mobile/MobileStore.java
package com.bendyline.gezel.mobile;

import com.bendyline.gezel.runtime.MobileModelStore;
import java.io.*;
import java.nio.charset.StandardCharsets;
import org.json.JSONObject;

final class MobileStore extends MobileModelStore {
    final ProductFiles productFiles;
    MobileStore(File filesDir) throws IOException { this(filesDir, true); }
    MobileStore(File filesDir, boolean recoverModels) throws IOException {
        super(new File(filesDir.getCanonicalFile(), "gezel"), recoverModels);
        productFiles = new ProductFiles(new File(root, "product"));
    }
    synchronized String readState() throws Exception {
        byte[] bytes = readDocument("state.json", MAX_STATE);
        if (bytes == null) return null;
        String text = decode(bytes);
        new JSONObject(text);
        return text;
    }

    synchronized void writeState(String text) throws Exception {
        byte[] bytes = text.getBytes(StandardCharsets.UTF_8);
        if (bytes.length > MAX_STATE) throw new IOException("State exceeds the 16 MB limit");
        new JSONObject(text);
        atomicWrite(new File(root, "state.json"), bytes);
    }

}

File: packages/mobile/native/ios/Sources/GezelMobileStorage/MobileStore.swift
import Foundation
#if canImport(GezelModelStorage)
@_exported import GezelModelStorage
#endif

public final class MobileStore: MobileModelStore, @unchecked Sendable {
    public let productFiles: ProductFiles
    private let root: URL
    private let fm = FileManager.default
    private let stateLock = NSLock()
    public override init(root: URL, recoverModels: Bool = true) throws {
        self.root = root
        self.productFiles = try ProductFiles(root: root.appendingPathComponent("product", isDirectory: true))
        try super.init(root: root, recoverModels: recoverModels)
    }
    private func withStateLock<T>(_ body: () throws -> T) rethrows -> T {
        stateLock.lock(); defer { stateLock.unlock() }
        return try body()
    }
    public func readState() throws -> String? {
        try withStateLock {
            let file = root.appendingPathComponent("state.json")
            guard fm.fileExists(atPath: file.path) else { return nil }
            let size = try file.resourceValues(forKeys: [.fileSizeKey]).fileSize ?? 0
            guard size <= Self.maximumStateBytes else { throw MobileStoreError.stateTooLarge }
            let handle = try FileHandle(forReadingFrom: file)
            defer { try? handle.close() }
            let data = try handle.read(upToCount: Self.maximumStateBytes + 1) ?? Data()
            try validateState(data)
            guard let text = String(data: data, encoding: .utf8) else { throw MobileStoreError.invalidState }
            return text
        }
    }

    public func writeState(_ text: String) throws {
        try withStateLock {
            let data = Data(text.utf8)
            try validateState(data)
            try data.write(to: root.appendingPathComponent("state.json"), options: .atomic)
        }
    }

    private func validateState(_ data: Data) throws {
        guard data.count <= Self.maximumStateBytes else { throw MobileStoreError.stateTooLarge }
        guard (try? JSONSerialization.jsonObject(with: data)) != nil else { throw MobileStoreError.invalidState }
    }

}
